### PR TITLE
Design doc for applying rules via matching

### PR DESCRIPTION
**Pull request by traiansf**
This should address the theory for solving Issue #940 .

Please be sure to check the Implementation concerns section at the bottom for relevance w.r.t. Issue #940.

---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

File: docs/2019-09-25-Applying-Axioms-By-Matching.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+`φ → •ψ`.
+
+QED
+
+
+Equality axioms
+---------------
+
+Given an axiom of the form 
+(0) `∀ Y . αₚ(Y) → αₜ(Y) = β(Y)`
```
Comment (virgil-serbanuta) [inline docs/2019-09-25-Applying-Axioms-By-Matching.md]: parenthesis, i.e. `αₚ(Y) → (αₜ(Y) = β(Y))`?

File: docs/2019-09-25-Applying-Axioms-By-Matching.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+- [Configuration Splitting Simplification](2018-11-08-Configuration-Splitting-Simplification.md)
+  and
+- [Applying axioms](2018-11-08-Applying-Axioms.md)
+
+whose proof of soundness assumes that both `φₜ` and `αₜ(Y)` are _function-like_
+and use the predicate `⌈φₜ ∧ αₜ(Y)⌉` to compute the substution and the remainder.
+
+Assuming that one or both of `φₜ` and `αₜ(Y)` are __not__ function-like,
+we cannot soundly apply the algorithm above.
```
Comment (virgil-serbanuta) [inline docs/2019-09-25-Applying-Axioms-By-Matching.md]: s/above/in that document/

File: docs/2019-09-25-Applying-Axioms-By-Matching.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+However,  if `⌈φₜ ∧ αₜ(Y)⌉` simplifies to a predicate of the form `(Y = t) ∧ p`
+such that `p` is `⊤` (unification did not introduce extra constraints)
+and if `t` verifies the conditions stated above, namely,
+
+- `φₜ = αₜ(t)` (syntactic, not semantic, equality) and
```
Comment (virgil-serbanuta) [inline docs/2019-09-25-Applying-Axioms-By-Matching.md]: Why syntactic? Is it just for optimization purposes? If yes, could you mention that?

File: docs/2019-09-25-Applying-Axioms-By-Matching.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+- `φₚ → αₚ(t)` (validity, not satisfiability)
```
Comment (ttuegel) [inline docs/2019-09-25-Applying-Axioms-By-Matching.md]: Does "validity" mean `⌊ ⌋`?

File: docs/2019-09-25-Applying-Axioms-By-Matching.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  and
+- [Applying axioms](2018-11-08-Applying-Axioms.md)
+
+whose proof of soundness assumes that both `φₜ` and `αₜ(Y)` are _function-like_
+and use the predicate `⌈φₜ ∧ αₜ(Y)⌉` to compute the substution and the remainder.
+
+Assuming that one or both of `φₜ` and `αₜ(Y)` are __not__ function-like,
+we cannot soundly apply the algorithm above.
+
+However,  if `⌈φₜ ∧ αₜ(Y)⌉` simplifies to a predicate of the form `(Y = t) ∧ p`
```
Comment (virgil-serbanuta) [inline docs/2019-09-25-Applying-Axioms-By-Matching.md]: What happens if the predicate is `(Y = t1) or (Y = t2)`? If that does not work, I'd add a comment. If that's too complicated for now, I'd do the same.

File: docs/2019-09-25-Applying-Axioms-By-Matching.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+such that `p` is `⊤` (unification did not introduce extra constraints)
```
Comment (virgil-serbanuta) [inline docs/2019-09-25-Applying-Axioms-By-Matching.md]: Why `top`?

File: docs/2019-09-25-Applying-Axioms-By-Matching.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+If `P` and `B` are equal, then the interpretation of `(φₜ = β(t))`
+is the whole set, whence the interpretation of `φₜ ∧ (φₜ = β(t))` is `P` and
+the interpretation of `β(t) ∧ (φₜ = β(t))` is `B`, thus they are equal.
+
+If `P` and `B` are not equal, then the interpretation of `(φₜ = β(t))`
+is the empty set, whence the interpretation of both `φₜ ∧ (φₜ = β(t))` and
+`β(t) ∧ (φₜ = β(t))` is the empty set, thus they are equal.
+
+
+Implementation concerns
```
Comment (ttuegel) [inline docs/2019-09-25-Applying-Axioms-By-Matching.md]: Do the constraints below imply that we cannot apply non-function-like simplification rules with side conditions?

File: docs/2019-09-25-Applying-Axioms-By-Matching.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+Rewrite rules
+-------------
+
+Given a rewrite axiom of the form 
+(0) `∀ Y .α(Y) → •β(Y)`
+and a goal `φ → •ψ` such that 
+(1) `φ` can be decomposed as `φ = φₜ ∧ φₚ`, (1') `φₚ` predicate.
+(2) `α(Y)` can be decomposed as `α(Y) = αₜ(Y) ∧ αₚ(Y)`
+and there exists `t` such that
```
[the remark this answers is not in the document]
Comment (yzhang90) [reply docs/2019-09-25-Applying-Axioms-By-Matching.md]: For all path, when you make a step, you can not look at only one rule, instead, you need to take the set of all semantics rules into account. As a result, the application of the above approach is quite limited.

**Comment by virgil-serbanuta** [comment]
> For all path, when you make a step, you can not look at only one rule ...

Yes, but this document was supposed to make us understand what we need to do with our code, and we're using the same code for all-path and one-path. If we should use this for non-functional stuff in one-path, and crash the program for all-path, then I'd like that to be obvious enough.

File: docs/2019-09-25-Applying-Axioms-By-Matching.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (ttuegel) [reply docs/2019-09-25-Applying-Axioms-By-Matching.md]: I don't think we care about applying non-function-like rewrite rules though, do we?

File: docs/2019-09-25-Applying-Axioms-By-Matching.md (hunk 4 above)
Comment (traiansf) [reply docs/2019-09-25-Applying-Axioms-By-Matching.md]: yes. i.e., not(φₚ → αₚ(t)) is unsat

File: docs/2019-09-25-Applying-Axioms-By-Matching.md (hunk 7 above)
Comment (traiansf) [reply docs/2019-09-25-Applying-Axioms-By-Matching.md]: we can only apply them if the top-level condition, which can be considered as part of phi_p implies the side condition.

File: docs/2019-09-25-Applying-Axioms-By-Matching.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+(3) `φₜ = αₜ(t)` and
+(4) `φₚ → αₚ(t)`
+
+Then, if we can show that
+(5) `β(t) ∧ φₚ → ψ`
```
[the remark this answers is not in the document]
Comment (traiansf) [reply docs/2019-09-25-Applying-Axioms-By-Matching.md]: yes.

File: docs/2019-09-25-Applying-Axioms-By-Matching.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (traiansf) [reply docs/2019-09-25-Applying-Axioms-By-Matching.md]: I also had this argument with Virgil. He suggested that if that's indeed the case, then we should make sure that (at least) for all-path this kind of rule application is only used for equalities.

File: docs/2019-09-25-Applying-Axioms-By-Matching.md (hunk 7 above)
Comment (ttuegel) [reply docs/2019-09-25-Applying-Axioms-By-Matching.md]: But that implication has to be _valid_ as described above, and we can't check that at the moment, so in practice that means no side conditions.

Comment (traiansf) [reply docs/2019-09-25-Applying-Axioms-By-Matching.md]: We can check validity by checking non-satisfiability of the negation.

File: docs/2019-09-25-Applying-Axioms-By-Matching.md (hunk 6 above)
Comment (traiansf) [reply docs/2019-09-25-Applying-Axioms-By-Matching.md]: I removed that by constraining the equality on being derivable from `φₚ`, whence if `φₚ` implies `p` it should be fine.

File: docs/2019-09-25-Applying-Axioms-By-Matching.md [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+Then it is true that `φ → •ψ`.
+
+### Proof
+
+
+Premise (0) says
+`∀ Y .α(Y) → •β(Y)`; instantiating `Y` with `t` we get
```
[the remark this answers is not in the document]
Comment (traiansf) [reply docs/2019-09-25-Applying-Axioms-By-Matching.md]: True, thanks for catching that!  I was thinking that we only build up substitutions with functional terms, but it should anyway be mentioned up-front and here where we use it, too.

File: docs/2019-09-25-Applying-Axioms-By-Matching.md [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+`β(t) ∧ φₚ → ψ`; by the Framing rule of ML with `•`, we get
+`•(β(t) ∧ φₚ) → •ψ`; by (7) and transitivity of `→`, we get
+`φ → •ψ`.
+
+QED
+
+### Coverage concerns (All-path-reachability)
+
+Note that if `φ` and `αₜ(Y)` are not constructor-based
+(in which case the unifier `Y = t` is an MGU) then
```
Comment (virgil-serbanuta) [inline docs/2019-09-25-Applying-Axioms-By-Matching.md]: As discussed:
* I think that it's hard to parse "in which case" as referring to "is constructor-based".
* I think that the constructors must be functional.
* Any injective symbol may work, not only constructors.

File: docs/2019-09-25-Applying-Axioms-By-Matching.md [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+it can safely be used when the configuration or the matching pattern
+are __not__ function-like.
+
+Rewrite rules
+-------------
+
+Given a rewrite axiom of the form 
+(0) `∀ Y .α(Y) → •β(Y)`
+and a goal `φ → •ψ` such that 
+(1) `φ` can be decomposed as `φ = φₜ ∧ φₚ`, (1') `φₚ` predicate.
```
[the remark this answers is not in the document]
Comment (traiansf) [reply docs/2019-09-25-Applying-Axioms-By-Matching.md]: I think most subscript characters are not portable.

File: docs/2019-09-25-Applying-Axioms-By-Matching.md [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+`φₜ ∧ φₚ → φₜ ∧ (φₜ = αₜ(t)) ∧ αₚ(t)`; by equality elimination we get
+`φₜ ∧ φₚ → αₜ(t) ∧ αₚ(t)`; by (6) and transitivity of `→`, we get
+`φₜ ∧ φₚ → •β(t)`; by FOL reasoning we get
+`φₜ ∧ φₚ → •β(t) ∧ φₚ`; due to (1') and properties of conjunction we get
+`φₜ ∧ φₚ → •(β(t) ∧ φₚ)`, which is equivalent by (1) with
+(7) `φ → •(β(t) ∧ φₚ)`.
+
+Premise (5) says
+`β(t) ∧ φₚ → ψ`; by the Framing rule of ML with `•`, we get
+`•(β(t) ∧ φₚ) → •ψ`; by (7) and transitivity of `→`, we get
```
[the remark this answers is not in the document]
Comment (traiansf) [reply docs/2019-09-25-Applying-Axioms-By-Matching.md]: Why? From: (7) `φ → •(β(t) ∧ φₚ)` and `•(β(t) ∧ φₚ) → •ψ` we deduce `φ → •ψ`

[the remark this answers is not in the document]
Comment (ttuegel) [reply docs/2019-09-25-Applying-Axioms-By-Matching.md]: Sorry, I should have said: I understand this step, but not what you mean by "transitivity". And now I understand both!
